### PR TITLE
COO GPU spmv kernel

### PR DESCRIPTION
**Pull request by yhmtsai**
COO GPU spmv kernel

### CODE REVIEW

**Comment by hartwiganzt** [comment]
Does the code run for you? I receive for the tester:

```./gpu/test/matrix/coo_kernels
Running main() from gtest_main.cc
[==========] Running 2 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 2 tests from Coo
[ RUN      ] Coo.SimpleApplyIsEquivalentToRef
unknown file: Failure
C++ exception with description "/Users/hanzt0114cl306/work/public_ginkgo/ginkgo/gpu/base/executor.cpp:118: raw_copy_to: cudaErrorLaunchFailure: unspecified launch failure" thrown in the test body.
/Users/hanzt0114cl306/work/public_ginkgo/ginkgo/gpu/test/matrix/coo_kernels.cpp:70: Failure
Expected: gpu->synchronize() doesn't throw an exception.
  Actual: it throws.
Unrecoverable CUDA error on device 0 in free: cudaErrorLaunchFailure: unspecified launch failure
Exiting program
```

**Comment by hartwiganzt** [comment]
works now - looks good to me!

**Comment by hartwiganzt** [comment]
@gflegar no conflicts here.

File: gpu/matrix/coo_kernels.cu [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+{
+    return atomicAdd(addr, val);
+}
+#endif
+
+
+__forceinline__ __device__ static cuDoubleComplex atomic_add(
+    cuDoubleComplex *address, cuDoubleComplex val)
+{
+    // Seperate to real part and imag part
```
Comment (pratikvn) [inline gpu/matrix/coo_kernels.cu]: nit: s/seperate/separate : here and a few places below.

File: gpu/matrix/coo_kernels.cu [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+        add_val = warp::shuffle_down(add_val, i);
+        if (threadIdx.x < 32 - i) {
+            *val += add_val;
+        }
+    }
+}
+
+
+template <typename ValueType, typename IndexType>
+__global__ __launch_bounds__(128) void spmv_kernel(
```
Comment (pratikvn) [inline gpu/matrix/coo_kernels.cu]: As this is the actual kernel, some documentation explaining the working would be nice.

File: gpu/matrix/coo_kernels.cu [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+
+}  // namespace
+
+
+template <typename ValueType, typename IndexType>
+void spmv(std::shared_ptr<const GpuExecutor> exec,
+          const matrix::Coo<ValueType, IndexType> *a,
+          const matrix::Dense<ValueType> *b, matrix::Dense<ValueType> *c)
+{
+    int multiple = 8;
```
Comment (pratikvn) [inline gpu/matrix/coo_kernels.cu]: Hard-coding values like this is always difficult to warrant. I guess here it is probably okay. But maybe it should always accompany some metrics/ reasoning as to why this value was chosen.

Comment (gflegar) [reply gpu/matrix/coo_kernels.cu]: I agree with you that it shouldn't look like this.

However, AFAIK, these are tuned parameters obtained from running experiments (and might differ for different architectures). I also have a similar problem in #49, and have them hardcoded.
The problem is that we don't have this problem resolved in Ginkgo - the idea is to be able to say something like:  "this is a parameter that should be tuned, valid values are from the set __S__, and use __x__ by default". Then the user should be able to do an autotuning run of ginkgo, which will set the parameters to proper values for their system.

However we do it right now, there's a good chance we'll have to change it when we implement autotuning, so I'm fine for now with having it like this + adding a "TODO" that this should be changed when we support autotuning.

**Comment by yhmtsai** [comment]
Thanks for your help.
I have fixed them.
@gflegar the complex version of atomic_add is not really atomic, but I think it is correct in COO spmv, right?

File: gpu/components/atomic.cuh [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+BIND_ATOMIC_ADD(unsigned int);
+BIND_ATOMIC_ADD(unsigned long long int);
+BIND_ATOMIC_ADD(float);
+
+
+#if (defined(CUDA_VERSION) && (CUDA_VERSION < 8000)) || \
+    (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ < 600))
+
+
+__forceinline__ __device__ static double atomic_add(double *addr, double val)
```
Comment (gflegar) [inline gpu/components/atomic.cuh]: this should also return `void`

**Comment by gflegar** [comment]
@yhmtsai yes, as it is written now (with functions returning void), the complex versions of atomic add are "atomic enough" for our use-case (atomic enough = if we have multiple calls to atomic add, they will all accumulate their values, even though it won't exactly be done atomically).

Everything looks good now. There's just this minor thing with the wrong return value in Kepler implementation of double atomic add. We can merge this as soon as it is fixed.

File: gpu/components/atomic.cuh (hunk 4 above)
Comment (gflegar) [reply gpu/components/atomic.cuh]: and no need for `static`

Comment (yhmtsai) [reply gpu/components/atomic.cuh]: fix it

**Pull request merged**